File: src/lib/pipeline/parsers/rico-ap-parser.ts
/**
 * rico-ap-parser.ts — Ricochet Agent Performance parser.
 *
 * Port of Python's rico_ap_parser.py to client-side TypeScript.
 *
 * Parses "Agent Performance (N).xlsx" files downloaded from Ricochet.
 * These contain pre-aggregated daily call counts per agent.
 *
 * The Python version reads the xlsx via zipfile + xml.etree due to an openpyxl
 * page-margin crash bug. SheetJS handles these files without issue, so we use
 * normal XLSX reading here.
 *
 * Expected columns:
 *   Name, Contacts, Real Time Calls, Live Queue Calls, Evening Queue Calls,
 *   Drip-Dial Calls, Boost Queue Calls, Email Queue Calls, Campaign Queue Calls,
 *   Direct Dial Calls, Inbound Calls, Total Outbound Calls, Total Queue Calls,
 *   Total Non Queue Calls, Total Calls, Total transfers in, Total transfers out
 *
 * Output rows contain:
 *   Agent, Calls, Inbound, Outbound
 *   (TalkTimeSeconds is NOT included — that comes from Rico CH)
 */

import * as XLSX from "xlsx"
import type { ParseResult } from "../types"
import { Spine } from "../spine"

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

/** Names to skip (dialers, summary rows) */
const SKIP_NAMES = new Set(["LM Dialer 1", "LM Dialer 2", "LM Dialer 3", "Total"])

/**
 * Header-to-field mapping. We try to match by header name first.
 * If headers don't match, we fall back to column index (0-indexed):
 *   A (0) = Name, K (10) = Inbound Calls, L (11) = Total Outbound Calls, O (14) = Total Calls
 */
const HEADER_MAP: Record<string, string> = {
  "Name": "Name",
  "Inbound Calls": "Inbound",
  "Total Outbound Calls": "Outbound",
  "Total Calls": "Calls",
}

/** Fallback column indices matching the Python _COL_MAP: A=0, K=10, L=11, O=14 */
const INDEX_FALLBACK: Record<number, string> = {
  0: "Name",
  10: "Inbound",
  11: "Outbound",
  14: "Calls",
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/** Safely convert a value to integer (mirrors Python _to_int) */
function toInt(val: unknown): number {
  if (val === null || val === undefined || val === "") return 0
  const n = Number(val)
  if (isNaN(n)) return 0
  return Math.floor(n)
}

/**
 * Try to build a column mapping from the header row.
 * Returns a map of column index → field name ("Name", "Inbound", "Outbound", "Calls")
 * if enough required headers are found, otherwise null (caller should use index fallback).
 */
function buildHeaderMap(headerRow: unknown[]): Map<number, string> | null {
  const map = new Map<number, string>()

  for (let i = 0; i < headerRow.length; i++) {
    const cellVal = String(headerRow[i] ?? "").trim()
    if (HEADER_MAP[cellVal]) {
      map.set(i, HEADER_MAP[cellVal])
    }
  }

  // We need at least "Name" and one numeric column to consider this valid
  const fields = new Set(map.values())
  if (fields.has("Name") && (fields.has("Calls") || fields.has("Inbound"))) {
    return map
  }
  return null
}

// ---------------------------------------------------------------------------
// Main parser
// ---------------------------------------------------------------------------

/**
 * Parse a Rico Agent Performance Excel file and return per-agent call counts.
 *
 * @param fileBuffer  - Raw file contents as ArrayBuffer
 * @param fileName    - Original filename (for logging)
 * @param spine       - Spine instance for agent name resolution
 * @param targetDate  - YYYY-MM-DD string for the date column, or null
 * @returns ParseResult with type "rico_ap"
 */
export function parseRicoAP(
  fileBuffer: ArrayBuffer,
  fileName: string,
  spine: Spine,
  targetDate: string | null,
): ParseResult {
  const logs: string[] = []
  logs.push(`[rico_ap] Parsing ${fileName}`)

  // ---------------------------------------------------------------
  // Read the workbook via SheetJS
  // ---------------------------------------------------------------
  let workbook: XLSX.WorkBook
  try {
    workbook = XLSX.read(fileBuffer, { type: "array" })
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err)
    logs.push(`[rico_ap] Error reading ${fileName}: ${msg}`)
    return { type: "rico_ap", rows: [], logs }
  }

  // Use the first sheet
  const sheetName = workbook.SheetNames[0]
  if (!sheetName) {
    logs.push(`[rico_ap] No sheets found in ${fileName}`)
    return { type: "rico_ap", rows: [], logs }
  }

  const sheet = workbook.Sheets[sheetName]
  // Convert to array-of-arrays (header: 1 gives raw rows without auto-keying)
  const rawRows: unknown[][] = XLSX.utils.sheet_to_json(sheet, {
    header: 1,
    defval: "",
  })

  if (rawRows.length < 2) {
    logs.push(`[rico_ap] No data rows in ${fileName}`)
    return { type: "rico_ap", rows: [], logs }
  }

  logs.push(`[rico_ap] Read ${rawRows.length} raw rows from sheet "${sheetName}"`)

  // ---------------------------------------------------------------
  // Determine column mapping (header-based or index fallback)
  // ---------------------------------------------------------------
  const headerRow = rawRows[0]
  const colMap = buildHeaderMap(headerRow)
  const useHeaderMap = colMap !== null

  if (useHeaderMap) {
    logs.push(`[rico_ap] Matched columns by header names`)
  } else {
    logs.push(`[rico_ap] Headers not recognized, using positional column fallback (A, K, L, O)`)
  }

  /**
   * Extract a named field from a row using either the header map or index fallback.
   */
  function getField(row: unknown[], field: string): unknown {
    if (useHeaderMap && colMap) {
      for (const [idx, f] of colMap.entries()) {
        if (f === field) return row[idx]
      }
      return ""
    }
    // Index fallback
    for (const [idx, f] of Object.entries(INDEX_FALLBACK)) {
      if (f === field) return row[Number(idx)]
    }
    return ""
  }

  // ---------------------------------------------------------------
  // Process data rows (skip header at index 0)
  // ---------------------------------------------------------------
  interface AgentAccum {
    calls: number
    inbound: number
    outbound: number
  }

  const agentMap = new Map<string, AgentAccum>()
  let skippedNames = 0
  let skippedResolve = 0

  for (let r = 1; r < rawRows.length; r++) {
    const row = rawRows[r]

    // Get name from column A / "Name" header
    const rawName = String(getField(row, "Name") ?? "").trim()
    if (!rawName) continue

    // Skip known non-agent names
    if (SKIP_NAMES.has(rawName)) {
      skippedNames++
      continue
    }

    // Resolve through spine
    const agent = spine.resolveAgent(rawName)
    if (!agent) {
      skippedResolve++
      continue
    }

    const calls = toInt(getField(row, "Calls"))
    const inbound = toInt(getField(row, "Inbound"))
    const outbound = toInt(getField(row, "Outbound"))

    // Aggregate in case of spine collisions (multiple rows → same canonical agent)
    const existing = agentMap.get(agent)
    if (existing) {
      existing.calls += calls
      existing.inbound += inbound
      existing.outbound += outbound
    } else {
      agentMap.set(agent, { calls, inbound, outbound })
    }
  }

  if (skippedNames > 0) {
    logs.push(`[rico_ap] Skipped ${skippedNames} non-agent rows (dialers/total)`)
  }
  if (skippedResolve > 0) {
    logs.push(`[rico_ap] Skipped ${skippedResolve} rows (unresolved agent name)`)
  }

  // ---------------------------------------------------------------
  // Build output rows
  // ---------------------------------------------------------------
  const rows: Record<string, unknown>[] = []
  for (const [agent, accum] of agentMap.entries()) {
    const row: Record<string, unknown> = {
      Agent: agent,
      Calls: accum.calls,
      Inbound: accum.inbound,
      Outbound: accum.outbound,
    }

    if (targetDate !== null) {
      row.Date = targetDate
    }

    rows.push(row)
  }

  logs.push(`[rico_ap] Parsed ${rows.length} agents from ${fileName}`)
  return { type: "rico_ap", rows, logs }
}

File: automation/src/parsers/rico_ap_parser.py
"""
rico_ap_parser.py — Ricochet Agent Performance parser.

Parses "Agent Performance (N).xlsx" files downloaded from Ricochet.
These contain pre-aggregated daily call counts per agent.

Note: These xlsx files cause openpyxl to crash due to a page-margin bug,
so we read them directly via zipfile + xml.etree.ElementTree.

Input: Agent Performance (N).xlsx with columns:
  Name, Contacts, Real Time Calls, Live Queue Calls, Evening Queue Calls,
  Drip-Dial Calls, Boost Queue Calls, Email Queue Calls, Campaign Queue Calls,
  Direct Dial Calls, Inbound Calls, Total Outbound Calls, Total Queue Calls,
  Total Non Queue Calls, Total Calls, Total transfers in, Total transfers out

Output: DataFrame with columns:
  Agent, Calls, Inbound, Outbound
  (TalkTimeSeconds is NOT included — that comes from Rico CH zips)
"""

import re
import zipfile
import xml.etree.ElementTree as ET
import pandas as pd
from pathlib import Path
from datetime import date as date_type, datetime, timedelta

from src.spine import Spine


# Column letters in the xlsx for the fields we need:
#   A = Name, K = Inbound Calls, L = Total Outbound Calls, O = Total Calls
_COL_MAP = {
    "A": "Name",
    "K": "Inbound",
    "L": "Outbound",
    "O": "Calls",
}

# Names to skip (dialers, owners, etc.)
_SKIP_NAMES = {"LM Dialer 1", "LM Dialer 2", "LM Dialer 3", "Total"}


def parse_downloads(
    downloads_folder: str,
    spine: Spine,
    target_date: date_type | None = None,
    batch_selector=None,
) -> pd.DataFrame:
    """
    Scan Downloads folder for Agent Performance xlsx files.

    Two modes of operation:

    1. Batch mode (batch_selector provided):
       Uses the pre-computed file assignment from BatchSelector.
       This handles Monday catch-ups where multiple files share the same
       download day. See batch_selector.py for the assignment algorithm.

    2. Single mode (no batch_selector):
       Finds a file whose mtime matches target_date + 1 (the expected
       download day for a normal weekday). Falls back to target_date
       mtime if the next-day file isn't found.
    """
    folder = Path(downloads_folder)

    # --- Batch mode: use pre-assigned file ---
    if batch_selector is not None:
        selected = batch_selector.get_file(target_date)
        if selected is None:
            print(f"[rico_ap_parser] No Agent Performance file assigned for {target_date}")
            return pd.DataFrame()
        print(f"[rico_ap_parser] Batch: using {selected.name} for {target_date}")
        return _parse_xlsx(selected, spine, target_date)

    # --- Single mode: find by mtime ---
    all_files = sorted(
        folder.glob("Agent Performance*.xlsx"),
        key=lambda f: f.stat().st_mtime,
        reverse=True,
    )

    if not all_files:
        print("[rico_ap_parser] No Agent Performance xlsx files found in Downloads")
        return pd.DataFrame()

    # Try to find a file downloaded the day after target_date
    selected = None
    if target_date is not None:
        next_day = target_date + timedelta(days=1)
        for f in all_files:
            mtime_date = datetime.fromtimestamp(f.stat().st_mtime).date()
            if mtime_date == next_day:
                selected = f
                break
        if selected is None:
            # Also check if there's a file from the target date itself
            for f in all_files:
                mtime_date = datetime.fromtimestamp(f.stat().st_mtime).date()
                if mtime_date == target_date:
                    selected = f
                    break

    if selected is None:
        # No file for the expected download date — don't silently use wrong data
        if target_date:
            print(
                f"[rico_ap_parser] No Agent Performance file for {target_date}. "
                f"Expected a file downloaded on {target_date + timedelta(days=1)} or {target_date}. "
                f"For weekend catch-ups, use --batch mode."
            )
            return pd.DataFrame()
        else:
            # No target date specified — just use newest
            selected = all_files[0]

    print(f"[rico_ap_parser] Using: {selected.name} (for target {target_date})")
    return _parse_xlsx(selected, spine, target_date)


def _parse_xlsx(
    file_path: Path,
    spine: Spine,
    target_date: date_type | None = None,
) -> pd.DataFrame:
    """Read Agent Performance xlsx via zipfile (bypasses openpyxl page-margin bug)."""
    try:
        strings = _read_shared_strings(file_path)
        rows_data = _read_sheet_rows(file_path, strings)
    except Exception as e:
        print(f"[rico_ap_parser] Error reading {file_path.name}: {e}")
        return pd.DataFrame()

    if not rows_data:
        return pd.DataFrame()

    # Skip header row (index 0)
    records = []
    for row in rows_data[1:]:
        name = row.get("A", "").strip()
        if not name or name in _SKIP_NAMES:
            continue

        # Resolve through spine
        agent = spine.resolve_agent(name)
        if not agent:
            continue

        calls = _to_int(row.get("O", "0"))
        inbound = _to_int(row.get("K", "0"))
        outbound = _to_int(row.get("L", "0"))

        records.append({
            "Agent": agent,
            "Calls": calls,
            "Inbound": inbound,
            "Outbound": outbound,
        })

    df = pd.DataFrame(records)

    if len(df) > 0:
        # Some agents may appear multiple times if spine maps collide; aggregate
        df = df.groupby("Agent").agg(
            Calls=("Calls", "sum"),
            Inbound=("Inbound", "sum"),
            Outbound=("Outbound", "sum"),
        ).reset_index()

        if target_date is not None:
            df["Date"] = target_date

    print(f"[rico_ap_parser] Parsed {len(df)} agents from {file_path.name}")
    return df


def _read_shared_strings(xlsx_path: Path) -> list[str]:
    """Read the shared strings table from an xlsx file."""
    ns = "http://schemas.openxmlformats.org/spreadsheetml/2006/main"
    strings = []
    with zipfile.ZipFile(xlsx_path) as zf:
        if "xl/sharedStrings.xml" not in zf.namelist():
            return strings
        tree = ET.parse(zf.open("xl/sharedStrings.xml"))
        for si in tree.findall(f".//{{{ns}}}si"):
            texts = si.findall(f".//{{{ns}}}t")
            strings.append("".join(t.text or "" for t in texts))
    return strings


def _read_sheet_rows(xlsx_path: Path, strings: list[str]) -> list[dict]:
    """Read all rows from sheet1, returning list of {col_letter: value} dicts."""
    ns = "http://schemas.openxmlformats.org/spreadsheetml/2006/main"
    rows_data = []
    with zipfile.ZipFile(xlsx_path) as zf:
        tree = ET.parse(zf.open("xl/worksheets/sheet1.xml"))
        for row in tree.findall(f".//{{{ns}}}row"):
            row_dict = {}
            for cell in row.findall(f"{{{ns}}}c"):
                ref = cell.get("r", "")
                cell_type = cell.get("t")
                v_elem = cell.find(f"{{{ns}}}v")
                val = v_elem.text if v_elem is not None else ""
                if cell_type == "s" and val:
                    val = strings[int(val)]
                # Extract column letter(s)
                col = "".join(ch for ch in ref if ch.isalpha())
                row_dict[col] = val
            rows_data.append(row_dict)
    return rows_data


def _to_int(val: str) -> int:
    """Safely convert a string to int."""
    try:
        return int(float(val))
    except (ValueError, TypeError):
        return 0
